File: bobROS/teensy_interface/src/simu.cpp
/*  
 * 
 * Copyright © 2022 DTU, 
 * Author:
 * Christian Andersen jcan@dtu.dk
 * 
 * The MIT License (MIT)  https://mit-license.org/
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the “Software”), to deal in the Software without restriction, 
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, 
 * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software 
 * is furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all copies 
 * or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR 
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE 
 * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, 
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN 
 * THE SOFTWARE. */

#include <string>
#include <string.h>
#include <inttypes.h>
#include "simu.h"
#include "steensy.h"
#include "uservice.h"
#include <stdlib.h>
#include "umqtt.h"
// create value
SImu imu[NUM_TEENSY_MAX];


void SImu::setup(int teensyNumber)
{ // ensure default values
  tn = teensyNumber;
  ini1 = "imu1teensy" + std::to_string(tn);
  ini2 = "imu2teensy" + std::to_string(tn);
  if (not ini.has(ini1))
  { // no data yet, so generate some default values
    // first IMU
    ini[ini1]["interval_gyro_ms"] = "12";
    ini[ini1]["interval_acc_ms"] = "12";
    ini[ini1]["gyro_offset"] = "0 0 0";
    ini[ini1]["log"] = "true";
    ini[ini1]["print_gyro"] = "false";
    ini[ini1]["print_acc"] = "false";
    // other IMU
    // ini[ini2]["use"] = "false";
    // ini[ini2]["interval_gyro_ms"] = "12";
    // ini[ini2]["interval_acc_ms"] = "12";
    // ini[ini2]["gyro_offset"] = "0 0 0";
    // ini[ini2]["log"] = "true";
    // ini[ini2]["print_gyro"] = "false";
    // ini[ini2]["print_acc"] = "false";
  }
  sub_gyro = strtol(ini[ini1]["interval_gyro_ms"].c_str(), nullptr, 10);
  sub_acc = strtol(ini[ini1]["interval_acc_ms"].c_str(), nullptr, 10);
  // use values and subscribe to source data
  // like teensy[x].send("sub pose 4\n");
  // std::string s = "sub gyro " + ini[ini1]["interval_gyro_ms"] + "\n";
  // teensy[tn].send(s.c_str());
  // s = "sub acc " + ini[ini1]["interval_acc_ms"] + "\n";
  // teensy[tn].send(s.c_str());
  // gyro offset
  const char * p1 = ini[ini1]["gyro_offset"].c_str();
  gyroOffset[0][0] = strtof(p1, (char**)&p1);
  gyroOffset[0][1] = strtof(p1, (char**)&p1);
  gyroOffset[0][2] = strtof(p1, (char**)&p1);
  // send calibration values to Teensy
  //
  toConsoleGyro[0] = ini[ini1]["print_gyro"] == "true";
  toConsoleAcc[0] = ini[ini1]["print_acc"] == "true";
  if (ini[ini1]["log"] == "true")
  { // open logfile
    std::string fn = service.logPath + "log_t" + to_string(tn) + "_gyro_1.txt";
    logfileGyro[0] = fopen(fn.c_str(), "w");
    fprintf(logfileGyro[0], "%% Gyro logfile (IMU1)\n");
    fprintf(logfileGyro[0], "%% 1 \tTime (sec)\n");
    fprintf(logfileGyro[0], "%% 2-4 \tGyro (x,y,z)\n");
    fprintf(logfileGyro[0], "%% Gyro offset %g %g %g\n", gyroOffset[0][0], gyroOffset[0][1], gyroOffset[0][2]);
    //
    fn = service.logPath + "log_t" + to_string(tn) + "_acc_1.txt";
    logfileAcc[0] = fopen(fn.c_str(), "w");
    fprintf(logfileAcc[0], "%% Accelerometer logfile (IMU1)\n");
    fprintf(logfileAcc[0], "%% 1 \tTime (sec)\n");
    fprintf(logfileAcc[0], "%% 2-4 \tAccelerometer (x,y,z)\n");
  }
  // other IMU
  if (ini[ini2]["use"] == "true")
  {
    // s = "sub gyro2 " + ini[ini2]["interval_gyro_ms"] + "\n";
    // teensy[tn].send(s.c_str());
    // s = "sub acc2 " + ini[ini2]["interval_acc_ms"] + "\n";
    // teensy[tn].send(s.c_str());
    // gyro offset
    p1 = ini[ini2]["gyro_offset"].c_str();
    gyroOffset[1][0] = strtof(p1, (char**)&p1);
    gyroOffset[1][1] = strtof(p1, (char**)&p1);
    gyroOffset[1][2] = strtof(p1, (char**)&p1);
    toConsoleGyro[1] = ini[ini2]["print_gyro"] == "true";
    toConsoleAcc[1] = ini[ini2]["print_acc"] == "true";
    if (ini[ini2]["log"] == "true" and logfileGyro[1] == nullptr)
    { // open logfile
      std::string fn = service.logPath + "log_t" + to_string(tn) + "_gyro_2.txt";
      logfileGyro[1] = fopen(fn.c_str(), "w");
      fprintf(logfileGyro[1], "%% Gyro logfile (IMU2)\n");
      fprintf(logfileGyro[1], "%% 1 \tTime (sec)\n");
      fprintf(logfileGyro[1], "%% 2-4 \tGyro (x,y,z)\n");
      fprintf(logfileGyro[1], "%% Gyro offset %g %g %g\n", gyroOffset[1][0], gyroOffset[1][1], gyroOffset[1][2]);
      //
      fn = service.logPath + "log_t" + to_string(tn) + "_acc_2.txt";
      logfileAcc[1] = fopen(fn.c_str(), "w");
      fprintf(logfileAcc[1], "%% Accelerometer logfile (IMU2)\n");
      fprintf(logfileAcc[1], "%% 1 \tTime (sec)\n");
      fprintf(logfileAcc[1], "%% 2-4 \tAccelerometer (x,y,z)\n");
    }
  }
}


void SImu::subscribeDataFromTeensy()
{
  if (sub_gyro > 0)
  {
    std::string s = "sub gyro " + to_string(sub_gyro) + "\n";
    teensy[tn].send(s.c_str());
  }
  if (sub_acc > 0)
  {
    std::string s = "sub acc " + to_string(sub_acc) + "\n";
    teensy[tn].send(s.c_str());
  }
  if (ini[ini2]["use"] == "true")
  {
    if (sub_gyro > 0)
    {
      std::string s = "sub gyro2 " + to_string(sub_gyro) + "\n";
      teensy[tn].send(s.c_str());
    }
    if (sub_acc > 0)
    {
      std::string s = "sub acc2 " + to_string(sub_acc) + "\n";
      teensy[tn].send(s.c_str());
    }
  }
  if (sub_gyro > 0 or sub_acc > 0)
  {
    service.logMessage("# SImu: subscribed to Teensy data (gyro and acc)");
    int m = 0;
    updTimeGyro[m].now();
    updTimeAcc[m].now();
  }
}

void SImu::terminate()
{
  for (int m = 0; m < 2; m++)
  {
    if (logfileAcc[m] != nullptr)
    {
      fclose(logfileAcc[m]);
      logfileAcc[m] = nullptr;
    }
    if (logfileGyro[m] != nullptr)
    {
      fclose(logfileGyro[m]);
      logfileGyro[m] = nullptr;
    }
  }
}

bool SImu::decode(const char* msg, UTime & msgTime)
{
  bool used = true;
  const char * p1 = msg;
  if (strncmp(p1, "acc ", 4) == 0)
  {
    if (strlen(p1) > 4)
      p1 += 4;
    else
      return false;
    float a[3];
    for (int i = 0; i < 3; i++)
      a[i] = strtof(p1, (char**)&p1);
    // IMU 1 (pt. one only)
    int m = 0;
    updTimeAcc[m] = msgTime;
    for (int i = 0; i < 3; i++)
      acc[m][i] = a[i];
    updateAccCnt[m]++;
    // save to log
    toLog(true, m);
  }
  else if (strncmp(p1, "gyro ", 5) == 0)
  {
    if (strlen(p1) > 5)
      p1 += 5;
    else
      return false;
    // get x,y and z values
    float g[3];
    for (int i = 0; i < 3; i++)
      g[i] = strtof(p1, (char**)&p1);
    // IMU number (there is one gyro only)
    int m = 0;
    updTimeGyro[m] = msgTime;
    for (int i = 0; i < 3; i++)
    {
      gyro[m][i] = g[i] - gyroOffset[m][i];
    }
    // notify users of a new update
    updateGyroCnt[m]++;
     // save to log (if requested)
    toLog(false, m);
    //
    if (inCalibration[m])
    { // Gyro calibration can be handled ambulant
      for (int j = 0; j < 3; j++)
        calibSum[m][j] = g[j];
      calibCount[m]++;
      printf("# gyro %d, %d : %g %g %g\n", m, calibCount[m], calibSum[m][0]/float(calibCount[m]), calibSum[m][1]/float(calibCount[m]), calibSum[m][2]/float(calibCount[m]));
      if (calibCount[m] >= calibCountMax)
      {
        for (int j = 0; j < 3; j++)
          gyroOffset[m][j] = calibSum[m][j]/calibCount[m];
        // implement new values
        const int MSL = 100;
        char s[MSL];
        snprintf(s, MSL, "%g %g %g", gyroOffset[m][0], gyroOffset[m][1], gyroOffset[m][2]);
        if (m == 0)
          ini[ini1]["gyro_offset"] = s;
        else
          ini[ini2]["gyro_offset"] = s;
        inCalibration[m] = false;
        //
        printf("# gyro %d calibration finished: %s\n", m, s);
      }
    }
  }
  else
    used = false;
  return used;
}

void SImu::toLog(bool accChanged, int imuIdx)
{
  if (service.stop)
    return;
  int m = imuIdx;
  if (accChanged)
  { // accelerometer
    if (logfileAcc[m] != nullptr and not service.stop_logging)
    {
      fprintf(logfileAcc[m],"%lu.%04ld %.4f %.4f %.4f\n", updTimeAcc[m].getSec(), updTimeAcc[m].getMicrosec()/100,
              acc[m][0], acc[m][1], acc[m][2]);
    }
    if (toConsoleAcc[m])
    {
      printf("%lu.%04ld %.4f %.4f %.4f Acc %d\n", updTimeAcc[m].getSec(), updTimeAcc[m].getMicrosec()/100,
             acc[m][0], acc[m][1], acc[m][2], imuIdx);
    }
  }
  else
  { // gyro data
    if (logfileGyro[m] != nullptr and not service.stop_logging)
    {
      fprintf(logfileGyro[m],"%lu.%04ld %.4f %.4f %.4f\n", updTimeGyro[m].getSec(), updTimeGyro[m].getMicrosec()/100,
              gyro[m][0], gyro[m][1], gyro[m][2]);
    }
    if (toConsoleGyro[m])
    {
      printf("%lu.%04ld %.4f %.4f %.4f Gyro %d\n", updTimeGyro[m].getSec(), updTimeGyro[m].getMicrosec()/100,
              gyro[m][0], gyro[m][1], gyro[m][2], imuIdx);
    }
  }
}

void SImu::calibrateGyro()
{ // calibrate both
  inCalibration[0] = true;
  inCalibration[1] = true;
}

void SImu::tick()
{
  int m = 0;
  // printf("# Imu tick, interval=%d time=%f\n", sub_gyro, updTimeGyro[m].getTimePassed());
  if ((sub_gyro > 0 and updTimeGyro[m].getTimePassed() > 2.5) or (sub_acc > 0 and updTimeAcc[m].getTimePassed() > 2.5))
  {
    subscribeDataFromTeensy();
  }
}

File: robobot/mqtt_python/simu.py
#/***************************************************************************
#*   Copyright (C) 2024 by DTU
#*   jcan@dtu.dk
#*
#*
#* The MIT License (MIT)  https://mit-license.org/
#*
#* Permission is hereby granted, free of charge, to any person obtaining a copy of this software
#* and associated documentation files (the “Software”), to deal in the Software without restriction,
#* including without limitation the rights to use, copy, modify, merge, publish, distribute,
#* sublicense, and/or sell copies of the Software, and to permit persons to whom the Software
#* is furnished to do so, subject to the following conditions:
#*
#* The above copyright notice and this permission notice shall be included in all copies
#* or substantial portions of the Software.
#*
#* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
#* INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
#* PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
#* FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
#* ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
#* THE SOFTWARE. */


import time as t
from datetime import *
import numpy as np
import aruco



class SImu:
    # AI
    aruco_client = None

    def _get_aruco_client(self):
      if self.aruco_client is None:
        self.aruco_client = aruco.get_client()
      return self.aruco_client

    def _get_aruco_pose(self):
      try:
        client = self._get_aruco_client()
        data = client.get_data()
        if not data or not data.get('estimates'):
          return None
        estimates = data['estimates']
        xs = [e['robot_x'] for e in estimates]
        ys = [e['robot_y'] for e in estimates]
        yaws = [e['robot_yaw'] for e in estimates]
        if len(estimates) == 0:
          return None
        avg_x = sum(xs) / len(xs)
        avg_y = sum(ys) / len(ys)
        avg_yaw = np.arctan2(np.sum(np.sin(yaws)), np.sum(np.cos(yaws)))
        return {'x': avg_x, 'y': avg_y, 'z': 0.0, 'yaw': avg_yaw}
      except Exception:
        return None
    ##
    gyro = [0, 0, 0]
    gyroUpdCnt = 0
    gyroTime = datetime.now()
    gyroInterval = 1

    acc  = [0, 0, 0]
    accTime = datetime.now()
    accUpdCnt = 0
    accInterval = 1

    def setup(self):
      # data subscription is set in teensy_interface/build/robot.ini
      from uservice import service
      loops = 0
      while not service.stop:
        t.sleep(0.01)
        if self.gyroUpdCnt == 0 or self.accUpdCnt == 0:
          # wait for data
          pass
        else: # finished
          print(f"% IMU (simu.py):: got data stream; {loops} loops.")
          break
        loops += 1
        if loops > 20:
          print(f"% IMU (simu.py):: no data updates after {loops} wait loops (continues).")
          break
        pass
        loops += 1
      # should we calibrate the gyro
      if service.args.gyro:
        print("% Starting calibrate gyro offset.")
        # ask Teensy to calibrate
        service.send("robobot/cmd/T0", "gyroc")
        # wait for calibration to finish (average over 1s)
        t.sleep(2.5)
        # save calibrated values
        service.send("robobot/cmd/T0", "eew")
        print("% Starting calibrate gyro offset finished.")
        t.sleep(0.5)
        # all done
        service.stop = True
      pass

    def print(self):
      from uservice import service
      print("% IMU acc  " + str(self.accTime - service.startTime) + " (" +
            str(self.acc[0]) + ", " +
            str(self.acc[1]) + ", " +
            str(self.acc[2]) + f") {self.gyroInterval:.4f} sec " +
            str(self.accUpdCnt))
      print("% IMU gyro " + str(self.gyroTime - service.startTime) + " (" +
            str(self.gyro[0]) + ", " +
            str(self.gyro[1]) + ", " +
            str(self.gyro[2]) + f") {self.accInterval:.4f} sec " +
            str(self.gyroUpdCnt))

    def decode(self, topic, msg):
        # decode MQTT message
        used = True
        if topic == "T0/gyro":
          gg = msg.split(" ")
          if (len(gg) >= 4):
            t0 = self.gyroTime;
            self.gyroTime = datetime.fromtimestamp(float(gg[0]))
            self.gyro[0] = float(gg[1])
            self.gyro[1] = float(gg[2])
            self.gyro[2] = float(gg[3])
            t1 = self.gyroTime;
            if self.gyroUpdCnt == 2:
              self.gyroInterval = (t1 -t0).total_seconds()
            else:
              self.gyroInterval = (self.gyroInterval * 99 + (t1 -t0).total_seconds()) / 100
            self.gyroUpdCnt += 1
            # self.print()

            from sfuse import iwo
            if iwo.wheelVelocityCnt > 0:
              dt = (t1 - t0).total_seconds()
              if 0 < dt:# < 0.5: # Handle if t0 lies in the past
                  if dt > 0.5:
                    print(f"% Warning: large dt={dt:.3f} sec between gyro updates!")
                  aruco_pose = self._get_aruco_pose()
                  pos_meas = None
                  att_meas = None
                  #if aruco_pose is not None:
                  #  pos_meas = [aruco_pose['x']/1000, aruco_pose['y']/1000, aruco_pose['z']/1000]
                  #  #pos_meas = [aruco_pose['x'], aruco_pose['y'], aruco_pose['z']]
                  #  att_meas = [np.deg2rad(aruco_pose['yaw'])]
                  iwo.fuse(self.acc, self.gyro, dt, pos_meas=pos_meas, att_meas=att_meas)
                  #print(f"Fused with dt={dt:.3f} sec, gyro={self.gyro}, acc={self.acc}, aruco_pose={aruco_pose}, fused_x={iwo.fused_x:.3f}, fused_y={iwo.fused_y:.3f}, fused_yaw={iwo.fused_yaw:.2f}")


        elif topic == "T0/acc":
          gg = msg.split(" ")
          if (len(gg) >= 4):
            t0 = self.accTime;
            self.accTime = datetime.fromtimestamp(float(gg[0]))
            self.acc[0] = float(gg[1])
            self.acc[1] = float(gg[2])
            self.acc[2] = float(gg[3])
            t1 = self.accTime;
            if self.accUpdCnt == 2:
               self.accInterval = (t1 -t0).total_seconds()
            else:
               self.accInterval = (self.accInterval * 99 + (t1 -t0).total_seconds()) / 100
            self.accUpdCnt += 1
            # self.print()
        else:
          used = False
        return used

    def terminate(self):
        print("% Pose terminated")
        pass

# create the data object
imu = SImu()

